### DISCUSSION TITLE
Flag to start next dev without cache

### DISCUSSION THREAD

**Discussion by gaearon** [vercel/next.js#33929] [Ideas]
### Describe the feature you'd like to request

I want to be able to make edits in `node_modules` and see the result without restarting the server. This is indispensable for debugging libraries, including Next.js itself.

### Describe the solution you'd like

```
yarn dev --no-cache
```

Build cache gets disabled. If I edit `node_modules`, I don't need to restart the server.

### Describe alternatives you've considered

What I do now:

```
rm -rf .next/cache
yarn dev
```

(make an edit)

```
rm -rf .next/cache
yarn dev
```

(make an edit)

```
rm -rf .next/cache
yarn dev
```

but this is not fun.

**Comment by migueloller** [comment]
To add some extra info to the discussion, this also makes it very hard to work with separate repos when using `npm link` or `yarn link` since Next.js sees those as being in `node_modules` and doesn't reload when changes are made.

**Comment by Robert-Schirmer** [comment]
I've been able to get hot refresh on changes to certain `node_module` with this `next.config.js`. 

By default all files in `node_modules` webpack assumes will not be modified. We need to update the `managedPaths` property of the webpack `snapshot` config to exclude whatever modules you want webpack to watch for changes in.
https://webpack.js.org/configuration/other-options/#managedpaths

```
let config = {
  reactStrictMode: true,
  swcMinify: true,
  eslint: {
    dirs: ['pages', 'src', '__tests__', 'cypress'],
  },
  webpack: (config) => {
    config.snapshot = {
      ...(config.snapshot ?? {}),
      // Add all node_modules but @next module to managedPaths
      // Allows for hot refresh of changes to @next module
      managedPaths: [/^(.+?[\\/]node_modules[\\/])(?!@next)/],
    };
    return config;
  },
};

if (process.env.ANALYZE === 'true') {
  const withBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: true,
  });
  config = withBundleAnalyzer(config);
}

module.exports = config;
```

**Reply by hirbod** [reply]
@Robert-Schirmer I want to give you the biggest hug in this world right now. You can't believe how annoyed I was with our `bit` managed modules. I owe you the biggest thank you!!!!

**Reply by byteab** [reply]
how would you do that with a monorepo

**Reply by alexandre-tobia** [reply]
@hirbod Hi, i'm working with bit too but i can't make the solution provided by @Robert-Schirme work. Could you share your config ? It would be awesome ! thanks in advance

**Reply by AdarshKonchady** [reply]
The above solution didn't work for me either. Wondering if it was working on a previous NextJs version.

**Reply by alexandre-tobia** [reply]
I've made it working with that next.config: 

```/** @type {import('next').NextConfig} */
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, nextRuntime, webpack }
  ) => {

    if(dev) {
      config.watchOptions = {
        followSymlinks: true,
      }

      config.snapshot.managedPaths = [];
      
    }

    return config;
  },
}
```

Not ideal in perf, but it works.

**Comment by paranat** [comment]
Just want to say that I'd really love to see this. Next engages in some aggressive caching of components, which can cause a lot of confusion during development. I would really like to see a solution for opting out of said caching during development without having to add code that will mess up the caching on production.

It can be an especially bad issue when switching between branches of a codebase.

**Comment by rusakovic** [comment]
Interesting that @gaearon himself wrote about this issue =)
Still in progress - it's strange from @vercel  side

**Comment by ApplePieGiraffe** [comment]
Any update on this? This is making locally developing a package to be consumed by a Next app difficult for us as you have to delete `.next/cache` and restart the server to see any changes you make to the package in the Next app

**Reply by nbkhope** [reply]
Having the same problem. Wish there was an ideal solution for this, not necessarily what this issue is asking for. Maybe related to https://github.com/vercel/next.js/discussions/57811

**Comment by elyselibetti** [comment]
Any updates on this? Caching on development feels like the polar opposite of what "development" is.

**Comment by mepadev** [comment]
I'm also experiencing a similar problem, but mine is related to **Fetch caching**. the cache doesn't work if I have the **Browser Console open** with the **'disable cache'** option enabled... It doesn't fail for all types of caches, but at least for request cases, it worked for me

**Reply by mepadev** [reply]
And I really wasn't aware of this condition, because if I close the console, the cache starts working again 🤣

**Comment by Christopher2K** [comment]
Coming back to this... would be nice to have something here....

**Comment by RichardODonoghue** [comment]
I like having a git hook which checks that the application is buildable when I commit code. Its infuriating that I have to restart my dev environment every time I commit because of Next borking the build cache each time.

**Comment by SuperRonanCraft** [comment]
Wild this is not a thing as of yet, especially as projects become larger and have aggressive back-end caching. Us front-end people  can't debug why a page hangs on first render then fine for X+ hours while the cache re-validates, so you can only test on every deployment cause the dev side takes too long to render to know if the page is taking a while cause of the page await or cause the dev environment takes like 10 seconds to render out again.
Would be nice to just disable caching entirely while keeping the "use cache" tags so we don't gotta go through each backend route, disabling the dozens of cached queries just to find that we needed one Suspense here and there.
Makes the user experience a hassle when a page takes a while to load after they click on a link and nothing happens while their data loads. While the developer is left scratching their head on why a route takes a while to load once and then fine until the cache becomes stale again...
